### PR TITLE
ARROW-3367: [INTEGRATION] Port Spark integration test to the docker-compose setup

### PR DESCRIPTION
**Pull request by kszucs**
This incorporates the fixes from https://github.com/apache/arrow/pull/3254
Running it on a local "Docker for Mac" takes ages because of its volume virtualization and the excessive IO usage of maven. So currently I need to run it on travis...

Crossbow builds [here](https://github.com/kszucs/crossbow/branches/all?utf8=%E2%9C%93&query=spark-integration)

### CODE REVIEW

File: integration/spark/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    M2_HOME=/root/.m2 \
+    PATH=/root/.m2/bin:/usr/local/maven/bin:$PATH
+RUN apt-get update -q -y && \
+    apt-get install -q -y openjdk-8-jdk && \
+    wget -q -O maven-$MAVEN_VERSION.tar.gz "https://www.apache.org/dyn/mirrors/mirrors.cgi?action=download&filename=maven/maven-3/$MAVEN_VERSION/binaries/apache-maven-$MAVEN_VERSION-bin.tar.gz" && \
+    tar -zxf /maven-$MAVEN_VERSION.tar.gz && \
+    rm /maven-$MAVEN_VERSION.tar.gz && \
+    mv /apache-maven-$MAVEN_VERSION /usr/local/maven
+
+# ARG SPARK_VERSION=2.4.0
```
Comment (xhochy) [inline integration/spark/Dockerfile]: Do you want to finish this here or will the spark download stay in the executed script?

**Review by xhochy — COMMENTED** [review]
LGTM except one small comment

File: integration/spark/Dockerfile (hunk 1 above)
Comment (kszucs) [reply integration/spark/Dockerfile]: The script uses a mounted directory to cache spark and its build artifacts. 
I was building spark in an image layer to boost things up on OSX, so this is just temporary.

File: integration/spark/runtest.sh [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# exit on any error
+set -e
+
+export MAVEN_OPTS="-q -Xmx2g -XX:ReservedCodeCacheSize=512m"
```
Comment (sarutak) [inline integration/spark/runtest.sh]: These options are passed to `java` command.
So if you suppose to suppress the output of `mvn` command, you should modify the mvn command bellow like as build/mvn -q`

Comment (kszucs) [reply integration/spark/runtest.sh]: Thanks :)

**Comment by wesm** [comment]
Is it possible to build the Java library inside the container rather than using the local volume?

**Comment by kszucs** [comment]
@wesm Sure, it's possible, but the two approaches serve different purposes.

I've optimized the current Dockerfiles to minimize `source code edit` <=> `build & test` cycles. If We want to build either Java or any other language implementation in the container, than the source must be `ADD`ed in a previous layer, and each code edits invalidates that layer, including the subsequent ones. 

So currently my approach is to build exclusively the dependencies in the container, than mount arrow's source to `/arrow` and another cache directory `/build` to host the build artifacts - while preventing the contamination of the source directory. This strategy speeds up the cpp and python builds a lot, and it doesn't require to both `build` and `run` the image.

We can optimize the other way around, build the libraries within the container, but that approach is more suited for production containers. In this particular case, We can build spark as a dependency in the container (image actually :)), but then We either need to build arrow before it (in a previous layer) or recompile afterwards hoping that maven can reuse the previous build artifacts. In the first case an edit would trigger a full rebuild.

Theoretically docker volumes don't have any performance penalty on linux hosts.

BTW it builds successfully on the DGX machine.

File: integration/spark/runtest.sh [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    "org.apache.spark.sql.execution.vectorized.ArrowColumnVectorSuite")
+
+  (echo "Testing Spark:"; IFS=$'\n'; echo "${SPARK_SCALA_TESTS[*]}")
+
+  # TODO: should be able to only build spark-sql tests with adding "-pl sql/core" but not currently working
+  build/mvn -Dtest=none -DwildcardSuites=$(IFS=,; echo "${SPARK_SCALA_TESTS[*]}") test
+
+  # Run pyarrow related Python tests only
+  echo "Testing PySpark:"
+  python/run-tests --modules pyspark-sql
```
Comment (kszucs) [inline integration/spark/runtest.sh]: [Testnames option](https://github.com/apache/arrow/pull/3300/commits/d6b42fb4bee6d1232df3b279d430e91f13c249bb#diff-15acc7899e9109bbb7e37a94ed4f20b6R103) is available after the 2.4.0 release.

**Comment by wesm** [comment]
```
$ docker-compose build spark-integration
Building spark-integration
Traceback (most recent call last):
  File "bin/docker-compose", line 6, in <module>
  File "compose/cli/main.py", line 71, in main
  File "compose/cli/main.py", line 127, in perform_command
  File "compose/cli/main.py", line 287, in build
  File "compose/project.py", line 384, in build
  File "compose/project.py", line 366, in build_service
  File "compose/service.py", line 1082, in build
  File "site-packages/docker/api/build.py", line 154, in build
  File "site-packages/docker/utils/build.py", line 30, in tar
  File "site-packages/docker/utils/build.py", line 49, in exclude_paths
  File "site-packages/docker/utils/build.py", line 214, in rec_walk
  File "site-packages/docker/utils/build.py", line 214, in rec_walk
  File "site-packages/docker/utils/build.py", line 214, in rec_walk
  [Previous line repeated 1 more time]
  File "site-packages/docker/utils/build.py", line 184, in rec_walk
PermissionError: [Errno 13] Permission denied: '/home/wesm/code/arrow/java/format/target/generated-sources'
[11337] Failed to execute script docker-compose
```

**Comment by kszucs** [comment]
I've never seen an error like that, the `docker-compose build` should not request access to `arrow/java`. Are You using `docker-compose.yml` in arrow's root?

**Comment by kszucs** [comment]
I got it now, docker can't gather the build context (arrow files).

File: docker-compose.yml [hunk 4]
```diff
@@ -21,13 +21,19 @@ version: '3.5'
...
+  - ubuntu-cache:/build:delegated
 
 x-alpine-volumes:
   &alpine-volumes
-  - .:/arrow:delegated
-  - ${ARROW_DOCKER_CACHE_DIR:-./docker_cache}/alpine:/build:delegated
+  - .:/arrow:ro  # ensures that docker won't contaminate the host directory
+  - alpine-cache:/build:delegated
+
+volumes:
```
Comment (kszucs) [inline docker-compose.yml]: Switched to named volumes.

File: docker-compose.yml [hunk 5]
```diff
@@ -21,13 +21,19 @@ version: '3.5'
 
 x-ubuntu-volumes:
   &ubuntu-volumes
-  - .:/arrow:delegated
-  - ${ARROW_DOCKER_CACHE_DIR:-./docker_cache}/ubuntu:/build:delegated
+  - .:/arrow:ro  # ensures that docker won't contaminate the host directory
```
Comment (kszucs) [inline docker-compose.yml]: This brought up new issues.

File: ci/docker_build_java.sh [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+# specific language governing permissions and limitations
+# under the License.
+
+set -e
+
+# /arrow/java is read-only
+mkdir -p /build/java
+rsync -a /arrow/header /build/java
+rsync -a /arrow/java /build/java
+rsync -a /arrow/format /build/java
```
Comment (kszucs) [inline ci/docker_build_java.sh]: I couldn't change maven's mind to contaminate the source directory, even with hacking the top-level pom.xml

File: docker-compose.yml (hunk 5 above)
Comment (wesm) [reply docker-compose.yml]: Hm, can you rsync the files into container space and build there?

Comment (kszucs) [reply docker-compose.yml]: Yes, but I don't like that solution, but probably have to stick with that.

**Comment by kszucs** [comment]
So finally I was able to retrieve the Arrow related issues of spark 1.4. A couple of test cases are failing (log is coming...)

**Comment by wesm** [comment]
@kszucs this needs a rebase. What is the status of this build?

**Comment by kszucs** [comment]
Docker didn't provide enough memory to compile spark, the error was not descriptive...

Pyspark tests are failing with:

```
Traceback (most recent call last):
  File "/spark/spark-2.4.0/python/pyspark/sql/tests.py", line 5937, in test_column_order
    grouped_df.apply(invalid_positional_types).collect()
AssertionError: "No cast implemented" does not match "An error occurred while calling o4388.collectToPython.
: org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 22.0 failed 1 times, most recent failure: Lost task 0.0 in stage 22.0 (TID 377, localhost, executor driver): org.apache.spark.api.python.PythonException:
 Traceback (most recent call last):
  File "/spark/spark-2.4.0/python/lib/pyspark.zip/pyspark/worker.py", line 372, in main
    process()
  File "/spark/spark-2.4.0/python/lib/pyspark.zip/pyspark/worker.py", line 367, in process
    serializer.dump_stream(func(split_index, iterator), outfile)
  File "/spark/spark-2.4.0/python/lib/pyspark.zip/pyspark/serializers.py", line 284, in dump_stream
    batch = _create_batch(series, self._timezone)
  File "/spark/spark-2.4.0/python/lib/pyspark.zip/pyspark/serializers.py", line 253, in _create_batch
    arrs = [create_array(s, t) for s, t in series]
  File "/spark/spark-2.4.0/python/lib/pyspark.zip/pyspark/serializers.py", line 253, in <listcomp>
    arrs = [create_array(s, t) for s, t in series]
  File "/spark/spark-2.4.0/python/lib/pyspark.zip/pyspark/serializers.py", line 251, in create_array
    return pa.Array.from_pandas(s, mask=mask, type=t)
  File "pyarrow/array.pxi", line 531, in pyarrow.lib.Array.from_pandas
  File "pyarrow/array.pxi", line 171, in pyarrow.lib.array
  File "pyarrow/array.pxi", line 80, in pyarrow.lib._ndarray_to_array
  File "pyarrow/error.pxi", line 91, in pyarrow.lib.check_status
pyarrow.lib.ArrowTypeError: an integer is required (got type str)
/arrow/cpp/src/arrow/python/helpers.cc:196 code: CheckPyError()
/arrow/cpp/src/arrow/python/python_to_arrow.cc:137 code: internal::CIntFromPython(obj, &value)
/arrow/cpp/src/arrow/python/iterators.h:54 code: func(objects[i], i, &keep_going)
/arrow/cpp/src/arrow/python/python_to_arrow.cc:1010 code: converter->AppendMultipleMasked(seq, mask, size)
```

**Comment by kszucs** [comment]
[This is the full log](https://gist.githubusercontent.com/kszucs/52843ed8b83429d0b65d94e6dd0774e4/raw/73920d648925935b4e03cedc403447ee8f43dfe6/spark-integration.log)

It'd be the best to merge this PR and decide/fix the issues in follow-up pull requests, mainly because this PR also helps to prevent contamination of source directory.

Please review @wesm @xhochy

File: docker-compose.yml [hunk 7]
```diff
@@ -105,8 +111,8 @@ services:
       context: .
       dockerfile: java/Dockerfile
     volumes:
-      - .:/arrow:delegated
-      - $HOME/.m2:/root/.m2:delegated
+      - .:/arrow:ro  # ensures that docker won't contaminate the host directory
```
Comment (kszucs) [inline docker-compose.yml]: https://issues.apache.org/jira/browse/ARROW-3078

**Comment by wesm** [comment]
Thanks @kszucs yes let's merge this and then look into the failure as follow up. @BryanCutler have you seent the above error?

File: ci/docker_build_python.sh [hunk 8]
```diff
@@ -26,15 +26,26 @@ export PYARROW_CMAKE_GENERATOR=Ninja
...
+  # directory is created by setuptools
+  mkdir -p $build_dir/pyarrow
 
-python setup.py build --build-temp=${build_dir} \
-                install --single-version-externally-managed \
-                        --record=/build/python/record.txt
+  relative_build_dir=$(realpath --relative-to=. $build_dir)
 
+  # this is a nightmare, but prevents mutating the source directory
+  # which is bind mounted as readonly
```
Comment (wesm) [inline ci/docker_build_python.sh]: oof sorry. We should report to setuptools about this issue since this is really crappy

**Review by wesm — APPROVED** [review]
+1, merging now. I'm sorry this was such a slog. Can you open a JIRA about the follow up fix?

**Comment by BryanCutler** [comment]
Spark updates for 0.12.0 patch here https://github.com/apache/spark/pull/23657

File: integration/spark/runtest.sh (hunk 3 above)
Comment (BryanCutler) [reply integration/spark/runtest.sh]: here is what I used to test pyarrow for all related python tests:
```
python/run-tests --testnames=pyspark.sql.tests.test_arrow,pyspark.sql.tests.test_pandas_udf,pyspark.sql.tests.test_pandas_udf_scalar,pyspark.sql.tests.test_pandas_udf_grouped_agg,pyspark.sql.tests.test_pandas_udf_grouped_map,pyspark.sql.tests.test_pandas_udf_window
```

Comment (BryanCutler) [reply integration/spark/runtest.sh]: @kszucs I just merged in support for 0.12.0 here https://github.com/apache/spark/commit/16990f929921b3f784a85f3afbe1a22fbe77d895, so you could also use the master branch

Comment (kszucs) [reply integration/spark/runtest.sh]: Thanks @BryanCutler !

**Pull request closed without merging**